Display custom run step properties in the lineage graph

labkey-klum opened the pull request:
#### Rationale
The Provenance API currently supports adding custom property values to a run step. The custom properties would be bound to an existing `Vocabulary` domain and while the server side code was persisting data to the database we did not surface them back to the lineage graph.

#### Related Pull Requests
https://github.com/LabKey/platform/pull/3177
https://github.com/LabKey/provenance/pull/97
https://github.com/LabKey/labkey-api-r/pull/76

#### Changes
- Modify the breadcrumb in the run step details grid to be : Back to Run Details
- Display more standard run step properties
- Display the optional custom properties with the standard run step properties in the run step details grid (include "Properties" column with set of model.detailColumns passed to DetailPanel, use CustomPropertiesRenderer via detailRenderer to render the key/value pairs for the step custom properties)

labkey-klum committed:
v2.145.1-fb-provenance-props.0

labkey-klum committed:
render run step custom properties
update breadcrumb for run step details view

labkey-klum committed:
merge from develop

cnathe commented on packages/components/src/internal/components/lineage/models.ts:
```diff
@@ -99,6 +99,16 @@ export interface ProvenanceMap {
     to: Experiment.LineageItemBase;
 }
 
+export interface RunStepProperties {
```
does this `RunStepProperties` interface need to be exported? I don't see any usages outside of this file

cnathe commented on packages/components/src/internal/components/lineage/node/LineageDetail.tsx:
```diff
@@ -13,6 +13,7 @@ export const LineageDetail: FC<LineageDetailProps> = memo(({ item }) => {
             baseFilters: item.pkFilters.map(pkFilter => Filter.create(pkFilter.fieldKey, pkFilter.value)),
             containerPath: item.container,
             schemaQuery: SchemaQuery.create(item.schemaName, item.queryName),
+            requiredColumns: ['*']
```
glad to see this worked

cnathe commented on packages/components/src/internal/components/lineage/node/LineageNodeDetail.tsx:
```diff
@@ -185,14 +185,17 @@ const RunStepNodeDetail: FC<RunStepNodeDetailProps> = memo(props => {
             <Tabs activeKey={tabKey} defaultActiveKey={1} id="lineage-run-step-tabs" onSelect={changeTab as any}>
                 <Tab eventKey={1} title="Step Details">
                     <LineageDetail item={step} />
+                    {step.properties.length > 0 && (
```
will the `step` object always have `properties` defined? or should this have a null check like `step?.properties.length`?

cnathe committed:
Merge branch 'develop' into fb_provenance_props

# Conflicts:
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
v2.148.0-fb-provenance-props.0

cnathe committed:
Display run graph step custom properties in main DetailPanel
- include "Properties" column with set of model.detailColumns passed to DetailPanel
- use CustomPropertiesRenderer via detailRenderer to render the key/value pairs for the step custom properties

cnathe committed:
v2.148.0-fb-provenance-props.1

cnathe committed:
Jest test for CustomPropertiesRenderer in LineageDetail.tsx

cnathe committed:
Remove unused RunStepProperties

cnathe committed:
v2.148.0-fb-provenance-props.2

labkey-klum commented:
Adding Nick as a reviewer as now most of the changes here are from @cnathe

labkey-nicka commented on packages/components/src/internal/components/lineage/node/LineageDetail.tsx:
```diff
@@ -1,22 +1,88 @@
+            <tbody>
+                {data.map(row => {
+                    const fieldKey = row.get('fieldKey');
+                    const name = fieldKey.substring(fieldKey.indexOf('#') + 1);
```
I guess I would expect the query metadata or queryInfo to give us more information so we don't have to parse the fieldKey to find out. I thought that Kevin had taken steps to add this to the query metadata returned from selectRows. If that is not made readily available it may be getting pulled through from the `QueryInfo` correctly. This would be the same column information you're filtering for above when you utilized the `ADDITIONAL_DETAIL_FIELDS`. The `col` should be made available via the render props.

cnathe replied:
The "Properties" column QueryColumn info we do have, but these fieldKeys for the key/value pairs within that properties value array are from the vocabulary domain. So I may be missing something here but I don't think we have further information about those vocab domain properties in the response object. Karl and I discussed adding in the name and caption/label to these response objects in the future but didn't plan to do that in this set of changes. @labkey-klum  or @labkey-nicka let me know if that vocab domain property info is there and I am just missing it.

labkey-nicka commented on packages/components/src/internal/components/lineage/node/LineageDetail.tsx:
```diff
@@ -1,22 +1,88 @@
 });
+
+function _resolveDetailRenderer(column: QueryColumn) {
+    let renderer = resolveDetailRenderer(column);
```
Hmm, makes me think we should consider falling back to `resolveDetailRenderer` and then forcing users to opt out of default rendering.

labkey-nicka commented on packages/components/src/internal/components/lineage/node/LineageDetail.tsx:
```diff
@@ -1,22 +1,88 @@
 
+const LineageDetailImpl: FC<LineageDetailProps & InjectedQueryModels> = memo(props => {
+    const { actions, queryModels } = props;
+    if (queryModels.model.isLoading) return <LoadingSpinner />;
```
Unfortunately, you'll need to handle all the problems here. Something like:

```tsx
if (queryModels.model.isLoading) return <LoadingSpinner />;
if (queryModels.model.hasLoadErrors) return <Alert>queryModels.model.loadErrors[0]</Alert>;
```

cnathe replied:
Thanks. Done.

cnathe committed:
CR feedback: handle model errors in LineageDetailImpl

cnathe committed:
v2.148.0-fb-provenance-props.3

labkey-klum committed:
merge from develop

labkey-klum committed:
update version and publish date

labkey-klum committed:
v2.149.3
